### PR TITLE
Datastore @Reference write support and integration tests

### PR DESCRIPTION
**Pull request by dmitry-s**
part of #1092

### CODE REVIEW

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java [hunk 1]
```diff
@@ -272,9 +274,40 @@ private Entity convertToEntityForSave(Object entity, Key... ancestors) {
...
+			}
+			Value<?> value;
+			if (persistentProperty.isCollectionLike()) {
+				if (val.getClass().isArray() && val.getClass() != byte[].class) {
+					//if a property is an array, convert it to list
+					val = CollectionUtils.arrayToList(val);
+				}
+				saveAll((Iterable<?>) val);
+				List<KeyValue> keyValues = StreamSupport.stream(((Iterable<?>) val).spliterator(), false)
+						.map(o -> KeyValue.of(this.getKey(o, false)))
```
Comment (ChengyuanZhao) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java]: Here we're not allocating new keys, but should we? I'm not thinking of any downsides right now to just allocating the keys for them, did you have some in mind?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java [hunk 2]
```diff
@@ -272,9 +274,40 @@ private Entity convertToEntityForSave(Object entity, Key... ancestors) {
...
+				.getPersistentEntity(entity.getClass());
+		datastorePersistentEntity.doWithReferenceProperties(persistentProperty -> {
+			PersistentPropertyAccessor accessor = datastorePersistentEntity.getPropertyAccessor(entity);
+			Object val = accessor.getProperty(persistentProperty);
+			if (val == null) {
+				return;
+			}
+			Value<?> value;
+			if (persistentProperty.isCollectionLike()) {
+				if (val.getClass().isArray() && val.getClass() != byte[].class) {
```
Comment (ChengyuanZhao) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java]: We're using this check so often  I think we should make this into a method for `DatastorePersistentProperty` called `isNonBytesArray`

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java (hunk 1 above)
Comment (dmitry-s) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java]: saveAll takes care of creating keys, so we don't need it to allocate them here

**Comment by dmitry-s** [comment]
@ChengyuanZhao PTAL

**Pull request merged**